tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,192 @@
+#
+#               std::vector<at::Dimname> names_vec;
+#               for (c10::IValue names_elem: names_list_in) {
+#                                ~~~~~~~~~~~~~~~~~~~~~~~~~ <-- Unrolling list, then convert elements one by one.
+#                   at::Dimname names_base = names_elem.to<at::Dimname>();
+#                   names_vec.push_back(names_base);
+#               }
+#               at::ArrayRef<at::Dimname> names_list_out(names_vec);
+#
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>(names_list_out);
+#           } else {
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+connector = "\n\t"
+
+
+# Convert all the arguments in a NativeFunction to C++ code, including TensorOptions.
+def convert_arguments(args: List[Argument]) -> Tuple[List[Expr], List[str]]:
+    argument_str = "c10::IValue {arg_name} = std::move(peek(stack, {pos}, {args_num}));"
+    expr_list = []
+    pop_ivalue = []
+    code_list = []
+    for i, arg in enumerate(args):
+        pop_ivalue.append(argument_str.format(arg_name=arg.name, pos=i, args_num=len(args)))
+        expr, code = argumenttype_ivalue_convert(arg.type, arg.name, mutable=arg.is_write)
+        expr_list.append(expr)
+        code_list.extend(code)
+    pop_ivalue.append("")
+    return expr_list, pop_ivalue + code_list
+
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, arg_name: str, *, mutable: bool = False) -> Tuple[Expr, List[str]]:
+    nctype = argumenttype_type(t=t, mutable=mutable, binds=arg_name)
+    ctype = nctype.cpp_type(strip_ref=True)
```

Comment:
In general we want to keep everything as a type in our type system for as long as possible, and only convert to string at the very end (for example when I see `_gen_code_optional_type` takes in a `string`, I have to scroll up here to see that it actually came from a `CType` object).

Can you just plumb around the `CType` directly, and call `.cpp_type()` on it directly in all of the f-strings where it gets used?

You can also plumb the underlying `CType` around instead of the `NamedCType`, since the name info is irrelevant in all of these sites. E.g. something like:
```
ctype: CType = cpp.argumenttype_type(t=t, mutable=mutable, binds=arg_name).type
```